homeassistant/components/application_credentials/__init__.py

```diff
@@ -0,0 +1,213 @@
+"""The Application Credentials integration.
+
+This integration provides APIs for managing local OAuth credentials on behalf of other
+integrations. The preferred approach for all OAuth integrations is to use the cloud
+account linking service, and this is the alternative for integrations that can't use
+it.
+
+Integrations register an authorization server, and then the APIs are used to add
+add one or more client credentials. Integrations may also provide credentials
+from yaml for backwards compatibility.
+"""
+from __future__ import annotations
+
+from dataclasses import dataclass
+import logging
+from typing import Protocol
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_CLIENT_ID, CONF_CLIENT_SECRET, CONF_DOMAIN, CONF_ID
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import collection, config_entry_oauth2_flow
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.storage import Store
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.loader import IntegrationNotFound, async_get_integration
+from homeassistant.util import slugify
+
+__all__ = ["ClientCredential", "AuthorizationServer", "async_import_client_credential"]
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "application_credentials"
+
+STORAGE_KEY = DOMAIN
+STORAGE_VERSION = 1
+DATA_STORAGE = "storage"
+
+CREATE_FIELDS = {
+    vol.Required(CONF_DOMAIN): cv.string,
+    vol.Required(CONF_CLIENT_ID): cv.string,
+    vol.Required(CONF_CLIENT_SECRET): cv.string,
+}
+UPDATE_FIELDS: dict = {}  # Not supported
+
+
+@dataclass
+class ClientCredential:
+    """Represent an OAuth client credential."""
+
+    client_id: str
+    client_secret: str
+
+
+@dataclass
+class AuthorizationServer:
+    """Represent an OAuth2 Authorization Server."""
+
+    authorize_url: str
+    token_url: str
+
+
+class ApplicationCredentialsStorageCollection(collection.StorageCollection):
+    """Application credential collection stored in storage."""
+
+    CREATE_SCHEMA = vol.Schema(CREATE_FIELDS)
+
+    async def _process_create_data(self, data: dict) -> dict:
+        """Validate the config is valid."""
+        result = self.CREATE_SCHEMA(data)
+        domain = result[CONF_DOMAIN]
+        if not await _get_platform(self.hass, domain):
+            raise ValueError("No application_credentials platform for %s" % domain)
+        return result
+
+    @callback
+    def _get_suggested_id(self, info: dict) -> str:
```

Comment:
Do we know the key type of the info dict?

Reply:
Updated typing.